### ISSUE TITLE
[Defactor] Vector combinators

### ISSUE THREAD

**Issue by femtomc** [genjax-community/genjax#232]
## Background

In the following discussion, I'll use `vector index` to refer to the address indices induced at the call site of a vector combinator. E.g. in `Gen.jl`, these are the integer-valued addresses which a vector combinator inserts into the address hierarchy.

The current implementations of the vector-shaped combinators are unsatisfying for a number of reasons.

### It's not easy to specify vector index updates

When constructing inference programs, it's often desirable to use a proposal with a custom address to target individual address indices (like the following):

```python
@genjax.gen
def proposal():
    x = Normal(0.0, 1.0) @ ("z", 1, "x")
```

Here, you could imagine using this to target the `("z", 1, "x")` address in a model program which supports a vector combinator call at `”z"`.

As of now, you can't use this proposal in importance sampling targeting a model program of the above form. The code path to support conversion from a hierarchical choice map to one which is compatible with `importance` on `Map` or `Unfold` is not supported.

### Access via vector index value is not supported

Currently, the semantics of vector shaped combinators in GenJAX is to take the inner generative function choice map and broadcast the leaf values (leaf, in JAX's sense -- e.g. a Pytree leaf).

The current implementations don't expose the vector index as an address index which a user can index with to slice along the struct of array representation.

I think it's beneficial to keep the struct of array representation, but support a "virtual" vector index. So e.g. the `VectorChoiceMap` looks like the following:

```
class VectorChoiceMap(ChoiceMap):
    inner: ChoiceMap
```

Inside of a `VectorChoiceMap`, the inner to choice map will still be broadcasted - but the user will be required to provide a toplevel virtual vector index to access sub-choices:

```
v_chm = VectorChoiceMap(...)
v = v_chm[1, "x"]
```

and the implementation will slice the leaf arrays using `jax.tree_util.tree_map` and then index into the result with `"x"`.

### `update` is inefficient

For both `Unfold` and `Map`, `update` walks the entire vector index list. 

Let's focus on `Map` for this discussion. Essentially, `Map` is like `jax.vmap` - in `update`, the `vmap` occurs simultaneously across the previous `VectorTrace`, and the new choice map. The implementation doesn't consider length or metadata associated with the choice map to decide how much to `vmap` (e.g. the minimal static update). Instead, it "optimizes" at runtime using masks and inserted switches.

I think we should consider a few common use cases to decide the best implementation strategy.

Let's assume that a user can specify a choice map update using either (ignore `EmptyChoiceMap`, which is trivial to support):

* `VectorChoiceMap` (which, by default, will require the user provide an inner choice map whose leaf lengths match the previous `VectorTrace` leaf lengths. This is a "full index" update.
* A new choice map called `IndexedChoiceMap` which carries _static index_ data and is the conversion target from proposals like the one described above. This allows users to specify subsets of indices to update. In general, the addresses allowed in proposals like the one above have to be static for the `BuiltinGenerativeFunction` language to accept them as valid addresses -- but the indices specified in `IndexedChoiceMap` can likely be dynamic/traced values.

These two code paths can be different implementations in the implementation of `importance` and `update` for `Map`. For `importance`, we have to visit everything anyways (there's potential for optimizations with `IndexedChoiceMap`, with two `vmap` calls -- then stacking on leaves and a permutation on leaf values).

For `update`, we again have two code paths. For `IndexedChoiceMap`, we should run the update _only on the length/number of indices in the `IndexedChoiceMap`. To support this, we'll likely need to mutate arrays on the leaves - we can benchmark this strategy versus a more complicated functional approach where we slice and stack. I suspect this will still be more efficient than running everything and inserting switches (as we would do if we just `vmap` over the entire previous `Trace`).

Now, the interface that the user should expect is that they can use proposals with address structure which specifies vector indices (e.g. `("x", 1, "z")`) -- so `BuiltinChoiceMap` that have addresses of this form for a vector-shaped call at `"x"` can be coerced into an `IndexedChoiceMap` as one of the code paths for `importance` and `update`.

### Constraints which differ across indices are not easily specified

This is an interesting one. Let's say a user uses a vector-shaped combinator, and wants to provide _different constraints_ on different vector indices.

This is currently possible, but largely impractical and would rely on internal `_src` functions - so there's no convenient public API to specify either an`IndexedChoiceMap` or a `VectorChoiceMap` to support this.

The problem here is the struct of array representation -- because the choice maps are SoA, different vector indices are just indexing into the leaf arrays -- so different choice maps at indices need to actually be represented by the same `Pytree`. The current mechanism to represent switching over structure is masking (which pushes the switches to runtime).

I think the right thing to do here is to support special logic in the constructors of `IndexedChoiceMap` and `VectorChoiceMap` that allows users to easily specify different shaped constraints on indices, and under the hood -- we do the appropriate masking management, etc.

**Comment by femtomc** [comment]
@midfield / @brianwa84 by the way, this issue deals with something I discussed with you both yesterday -- what's the right way to communicate vector-shaped constraints to the vector combinators (`MapCombinator` and `UnfoldCombinator`).

This is what I'll be working on during my coding time tomorrow, so if you happen to read and have opinions -- I'd welcome them during the next few days.

**Comment by femtomc** [comment]
@derifatives (context -- this is what I was discussing with you + Ben today)

**Comment by femtomc** [comment]
Closed by #413.
